### DISCUSSION TITLE
Render a full page component through a controller without adding a view

### DISCUSSION THREAD

**Discussion by mokhosh** [livewire/livewire#2255] [1. Help] [answered]
TLDR; Can you return or render a full page Livewire component from a controller without adding another blade view?

Context: I have two models in my app, both of which have slugs. I want to catch example.com/{slug} urls, see if an instance of either of those models exists with that slug and render the corresponding full page livewire component. I can't use the same component for both, because they are very different.

**Comment by mokhosh** [comment]
This is the closest I've got, but this doesn't include the default layout:
```php
return Livewire::mount('project-page', ['project' => $model])->html();
```

**Reply by shanerbaner82** [reply]
Still works in the future!

**Comment by mokhosh** [comment]
Ok this took another step. This renders the whole page, but I can't pass route parameters and make the model binding happen:
```php
return (new ProjectPage)(app(), $request->route());
```

**Comment by mokhosh** [comment, marked as answer]
Ok, got it working by adding this line:
```php
$request->route()->setParameter('project', $model->id);
```

Now everything works.
But I'm curious is there a more elegant way of doing this?

**Reply by emmaLaraDev** [reply]
@mokhosh I'm trying to do the same as you did, but could you share your Controller? or could you be more specific? I'm new on this :)

**Comment by s-patompong** [comment]
This is how I did it in my project.

This is inside my controller. In this case, `PaymentSuccess` is a full-page Livewire component.

```php
public function __invoke(string $paymentToken)
{
    $paymentProcess = new PaymentSuccess();

    // Assign any public parameter here
    $paymentProcess->token = $paymentToken;

    return App::call([$paymentProcess, '__invoke']);
}
```

As you can see in the code, you can assign the public parameter via the instance of the component class. Then, we use App::call to call the `__invoke` method (Laravel will do DI in the invoke method automatically).

**Reply by JanneDeVos** [reply]
I tried your solution but my token appears null. Is their something I'm doing wrong?

**Reply by mokhosh** [reply]
@DuRenard my method has been working fine if you want to give it a try

**Reply by JanneDeVos** [reply]
@mokhosh I'm using your solution in my project and it's working just fine. 

Just like you already asked, I'm always looking for the most elegant way of doing things aswel. In my opinion, the solution from @s-patompong is looking a bit more elegant to me. So I just asked out of curiosity.

**Reply by s-patompong** [reply]
> I tried your solution but my token appears null. Is their something I'm doing wrong?

@DuRenard Can I see your Livewire component?

You need to make sure that the token is a public property on the component :)

**Comment by RizwanNaasir** [comment]
If anybody still having issue with rendering the Component from Controller this is how I am rendering also passing params to Public Property of Component Class.

```
use App\Http\Livewire\UserRegister;
---
public function form(Request $request)
    {
        $route = $request->route();
        $route->setParameter('user',auth()->user());
        return (new UserRegister)(app(),$route);
    }
---
```

**Comment by ghzb** [comment]
The base class`Livewire\Component` has an `__invoke` method and seems to use dependency injection in the signature. Doing this in your controller should render the component as a full page.

```php
use Illuminate\Support\Facades\App;

class MyController extends Controller 
{
    public function index()
    {
         return App::call(MyLivewireComponent::class);
    }
}
```

**Reply by devmount** [reply]
Thanks! This should be the accepted answer.
